File: data_environment/sources/scope_reduction.1.f90
! @@name:	scope_reduction.1f
! @@type:	F-free
! @@compilable:	yes
! @@linkable:	no
! @@expect:	success
! @@version:	omp_5.1
subroutine do_work(n, a, s)
   implicit none
   integer n, i
   real a(*), s, loc_s
   integer, save :: nthrs

   loc_s = 0.0                ! local sum
   !$omp do
      do i = 1, n
         loc_s = loc_s + a(i)
      end do
   !$omp single
      s = 0.0                 ! total sum
      nthrs = 0
   !$omp end single
   !$omp scope reduction(+:s,nthrs)
      s = s + loc_s
      nthrs = nthrs + 1
   !$omp end scope
   !$omp masked
      print *, "total sum = ", s, ", nthrs = ", nthrs
   !$omp end masked
end subroutine

function work(n, a) result(s)
   implicit none
   integer n
   real a(*), s

   !$omp parallel
      call do_work(n, a, s)
   !$omp end parallel
end function

File: data_environment/sources/scope_reduction.1.cpp
/*
* @@name:	scope_reduction.1
* @@type:	C++
* @@operation:	compile
* @@expect:	success
* @@version:	omp_5.1
*/
#include <stdio.h>
void do_work(int n, float a[], float &s)
{
   float loc_s = 0.0f;        // local sum
   static int nthrs;
   #pragma omp for
      for (int i = 0; i < n; i++)
         loc_s += a[i];
   #pragma omp single
   {
      s = 0.0f;               // total sum
      nthrs = 0;
   }
   #pragma omp scope reduction(+:s,nthrs)
   {
      s += loc_s;
      nthrs++;
   }
   #pragma omp masked
      printf("total sum = %f, nthrs = %d\n", s, nthrs);
}

float work(int n, float a[])
{
   float s;
   #pragma omp parallel
   {
      do_work(n, a, s);
   }
   return s;
}
